web/modules/custom/state_machine_permissions/src/StateMachinePermissions.php

```diff
@@ -0,0 +1,105 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Drupal\state_machine_permissions;
+
+use Drupal\Core\DependencyInjection\ContainerInjectionInterface;
+use Drupal\Core\StringTranslation\StringTranslationTrait;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Dynamic permissions provider.
+ */
+class StateMachinePermissions implements ContainerInjectionInterface {
+
+  use StringTranslationTrait;
+
+  /**
+   * The permission helper service.
+   *
+   * @var \Drupal\state_machine_permissions\StateMachinePermissionsHelperInterface
+   */
+  protected $permissionsHelper;
+
+  /**
+   * Constructs a state machine permissions object.
+   *
+   * @param \Drupal\state_machine_permissions\StateMachinePermissionsHelperInterface $permissionsHelper
+   *   The permission helper service.
+   */
+  public function __construct(StateMachinePermissionsHelperInterface $permissionsHelper) {
+    $this->permissionsHelper = $permissionsHelper;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    return new static($container->get('state_machine_permissions.helper'));
+  }
+
+  /**
+   * Returns an array of permissions related to state machine.
+   *
+   * @return array
+   *   Dynamically generated permissions.
+   */
+  public function stateMachinePermissions(): array {
+    $permissions = [];
+    $state_field_map = $this->permissionsHelper->getStateFieldMap();
+    if (empty($state_field_map)) {
+      return $permissions;
+    }
+
+    foreach ($state_field_map as $entity_type_id => $bundles) {
+      foreach ($bundles as $bundle => $fields) {
+        $permissions += $this->generateWorkflowPermissionsForBundle($entity_type_id, $bundle, $fields);
+      }
+    }
+    return $permissions;
+  }
+
+  /**
+   * Returns an array of permissions per transmission for the given bundle.
+   *
+   * @param string $entity_type_id
+   *   The entity type id.
+   * @param string $bundle
+   *   The entity bundle.
+   * @param array $field_names
+   *   An array of state field names.
+   *
+   * @return array
+   *   An array of permissions keyed by the permission id where each entry
+   *   contains the 'label' key and the description.
+   */
+  public function generateWorkflowPermissionsForBundle($entity_type_id, $bundle, array $field_names) {
```

Comment:
Since this method is not called anywhere outside of the scope of this class, this can be a protected method. Let's also add full type hint information.